README.md

```diff
@@ -12,51 +12,135 @@ decrypted only by the controller running in the target cluster and
+
+In particular, the annotations and labels of a `SealedSecret` resource, are not the same as the annotations of the `Secret` that gets generated out of it.
+
+To capture this distinction, the `SealedSecret` object has a `template` section which encodes all the fields want the controller uses when producing the unsealed `Secret`.
+
+This includes metadata such as labels or annotations, but also things like the `type` of the secret.
+
+```yaml
+apiVersion: bitnami.com/v1alpha1
+kind: SealedSecret
+metadata:
+  name: mysecret
+  namespace: mynamespace
+  annotation:
+    "kubectl.kubernetes.io/last-applied-configuration": ....
+spec:
+  encryptedData:
+    .dockercfg: AgBy3i4OJSWK+PiTySYZZA9rO43cGDEq.....
+  template:
+    type: kubernetes.io/dockercfg
+    # this is an example of labels and annotations that will be added to the output secret
+    metadata:
+      labels:
+        "jenkins.io/credentials-type": usernamePassword
+      annotations:
+        "jenkins.io/credentials-description": credentials from Kubernetes
+```
+
+The controller would unseal that into something like:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: mysecret
+  namespace: mynamespace
+  labels:
+    "jenkins.io/credentials-type": usernamePassword
+  annotations:
+    "jenkins.io/credentials-description": credentials from Kubernetes
+  ownerReferences:
+  - apiVersion: bitnami.com/v1alpha1
+    controller: true
+    kind: SealedSecret
+    name: mysecret
+    uid: 5caff6a0-c9ac-11e9-881e-42010aac003e
+type: kubernetes.io/dockercfg
+data:
+  .dockercfg: ewogICJjcmVk...
+```
+
+As you can see, the generated `Secret` resource is a "dependent object" of the `SealedSecret` and as such
+it will be updated and deleted whenever the `SealedSecret` object gets updated or deleted.
+
+
+### Public key / Certificate
 
 The key certificate (public key portion) is used for sealing secrets,
 and needs to be available wherever `kubeseal` is going to be
 used. The certificate is not secret information, although you need to
-ensure you are using the correct file.
+ensure you are using the correct one.
 
 `kubeseal` will fetch the certificate from the controller at runtime
 (requires secure access to the Kubernetes API server), which is
-convenient for interactive use.  The recommended automation workflow
-is to store the certificate to local disk with
+convenient for interactive use, but it's known to be brittle when users
+have clusters with special configurations such *private GKE clusters* that have
+firewalls between master and nodes.
+
+An alternative workflow
+is to store the certificate somewhere (e.g. local disk) with
 `kubeseal --fetch-cert >mycert.pem`,
 and use it offline with `kubeseal --cert mycert.pem`.
 The certificate is also printed to the controller log on startup.
 
+
+> **NOTE**: we are working on providing key management mechanisms that offload the encryption to HSM based modules or managed cloud crypto solutions such as KMS.
+
+## Installation
```

Comment:
I can't decide if this should come just after the overview or not. The "SealedSecrets as templates for secrets" and "Public key / Certificate" sections are great info but more about how to use it, and perhaps not quite as important as how to install it in the first place.

What do you think?

Reply:
I flipped those sections around a few times, unable to make up my mind about that.

I ended up with this order because of these reasons:

1. The installation from sources is long-ish and would unnecessarily deflect attention. On the other hand I didn't find a good place to bury it; it's kinda/sorta related to the Development section, but not necessarily. There are people who like to build stuff from sources even if they are not really doing that as contributors and thus it might not be fair to subject them to the full cost of a development environment (which admittedly in this case is just `make`, but it might change). Similar situation for homebrew, helm chart etc

2. It seems more logical to me to first let users know what this tool is and whether they actually want to use it, and once they decided they want to give it a try it should be easy for them to do it, but I don't feel comfortable imposing lengthly installation instructions users would need to skip over just to get a feeling of what this thing is.

3. The actual installation instructions are in the release page, so that they can be directly copy&paste a link with the full version number instead of having to do

But your feedback makes perfect sense too. What if we just link to the Install section as early as possible, and let the readers themselves choose when they want us to shut up and take the money?

Eventually I wanted to split the verbose parts in separate docs so perhaps we'd have to have a link to it anyway.

Reply:
An alternative view ...

Someone tells me I should use SealedSecrets because it does exactly what I want - makes my k8s secret actually secret :). I go to the github page, read the overview and am satisfied I reached the right place (the overview). All I want to do next is install it and start playing, but there's a load of other text to skip over first.

But I'm not sure either. Feel free to land as-is, it's just a suggestion.